src/baseTrie.js

```diff
@@ -24,752 +22,759 @@ module.exports = Trie
- * @memberof Trie
- * @param {Buffer|String} key - the key to search for
- * @param {Function} cb A callback `Function` which is given the arguments `err` - for errors that may have occured and `value` - the found value in a `Buffer` or if no value was found `null`
- */
-Trie.prototype.get = function (key, cb) {
-  var self = this
-
-  key = ethUtil.toBuffer(key)
-
-  self.findPath(key, function (err, node, remainder, stack) {
-    var value = null
-    if (node && remainder.length === 0) {
-      value = node.value
-    }
-
-    cb(err, value)
-  })
-}
+module.exports = class Trie {
+  constructor (db, root) {
+    const self = this
+    this.EMPTY_TRIE_ROOT = ethUtil.SHA3_RLP
+    this.sem = semaphore(1)
+
+    // setup dbs
+    this.db = db || level()
+
+    this._getDBs = [this.db]
+    this._putDBs = [this.db]
+
+    Object.defineProperty(this, 'root', {
+      set(value) {
+        if (value) {
+          value = ethUtil.toBuffer(value)
+          assert(value.length === 32, 'Invalid root length. Roots are 32 bytes')
+        } else {
+          value = self.EMPTY_TRIE_ROOT
+        }
 
-/**
- * Stores a given `value` at the given `key`
- * @method put
- * @memberof Trie
- * @param {Buffer|String} key
- * @param {Buffer|String} Value
- * @param {Function} cb A callback `Function` which is given the argument `err` - for errors that may have occured
- */
-Trie.prototype.put = function (key, value, cb) {
-  var self = this
-
-  key = ethUtil.toBuffer(key)
-  value = ethUtil.toBuffer(value)
-
-  if (!value || value.toString() === '') {
-    self.del(key, cb)
-  } else {
-    cb = callTogether(cb, self.sem.leave)
-
-    self.sem.take(function () {
-      if (self.root.toString('hex') !== ethUtil.SHA3_RLP.toString('hex')) {
-        // first try to find the give key or its nearst node
-        self.findPath(key, function (err, foundValue, keyRemainder, stack) {
-          if (err) {
-            return cb(err)
-          }
-          // then update
-          self._updateNode(key, value, keyRemainder, stack, cb)
-        })
-      } else {
-        self._createInitialNode(key, value, cb) // if no root initialize this trie
+        this._root = value
+      },
+      get() {
+        return this._root
       }
     })
-  }
-}
 
-/**
- * deletes a value given a `key`
- * @method del
- * @memberof Trie
- * @param {Buffer|String} key
- * @param {Function} callback the callback `Function`
- */
-Trie.prototype.del = function (key, cb) {
-  var self = this
-
-  key = ethUtil.toBuffer(key)
-  cb = callTogether(cb, self.sem.leave)
+    this.root = root
+
+    /**
+     * Writes a value directly to the underlining db
+     * @method putRaw
+     * @memberof Trie
+     * @param {Buffer|String} key The key as a `Buffer` or `String`
+     * @param {Buffer} value The value to be stored
+     * @param {Function} callback A callback `Function`, which is given the argument `err` - for errors that may have occured
+     */
+    this.putRaw = this._putRaw
```

Comment:
This move of ``putRaw`` to here puts public API methods into a strange/unordered order in the docs and it is also not so good for code readability to have this proxy method.

Do you have an overview how much effort it would be to remove the proxy method and would it eventually be worth the extra effort? An short-run idea for the documentation order would be to move this JSDoc part down to the ``_putRaw`` method (and keep the ``putRaw`` naming in the JSDoc comment), a bit cheating but maybe ok in this context.

Reply:
Moving the JSDoc part down seems to be a good short term solution. I tried rewriting `putRaw` as 
```
  putRaw (...args) {
    return this._putRaw(...args)
  }
```
but ran into `RangeError: Maximum call stack size exceeded` issues in tests due to `checkpoint-interface` (some methods seem to be getting swapped there). Once I figure out what's going on there we may be able to remove the proxy method.